components/brave_today/browser/direct_feed_controller.cc

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// you can obtain one at http://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_today/browser/direct_feed_controller.h"
+
+#include <algorithm>
+#include <codecvt>
+#include <iterator>
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "base/barrier_callback.h"
+#include "base/callback.h"
+#include "base/containers/flat_set.h"
+#include "base/logging.h"
+#include "base/time/time.h"
+#include "brave/components/brave_private_cdn/headers.h"
+#include "brave/components/brave_today/browser/network.h"
+#include "brave/components/brave_today/common/brave_news.mojom-forward.h"
+#include "brave/components/brave_today/common/brave_news.mojom-shared.h"
+#include "brave/components/brave_today/common/brave_news.mojom.h"
+#include "brave/components/brave_today/common/pref_names.h"
+#include "brave/components/brave_today/rust/lib.rs.h"
+#include "components/prefs/pref_service.h"
+#include "net/base/load_flags.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "services/network/public/cpp/shared_url_loader_factory.h"
+#include "services/network/public/cpp/simple_url_loader.h"
+#include "services/network/public/mojom/url_response_head.mojom.h"
+#include "ui/base/l10n/time_format.h"
+
+namespace brave_news {
+
+namespace {
+
+mojom::ArticlePtr RustFeedItemToArticle(const FeedItem& rust_feed_item) {
+  auto metadata = mojom::FeedItemMetadata::New();
+  metadata->title = static_cast<std::string>(rust_feed_item.title);
+  metadata->image = mojom::Image::NewImageUrl(
+      GURL(static_cast<std::string>(rust_feed_item.image_url)));
+  metadata->url =
+      GURL(static_cast<std::string>(rust_feed_item.destionation_url));
+  metadata->description = static_cast<std::string>(rust_feed_item.description);
```

Comment:
Hmm, I actually don't want to strip html tags. We parse them later (in JS) to check for an image: https://github.com/brave/brave-core/blob/f41b32de131c13a7dfb1144b795290b1f0e5dce1/components/brave_new_tab_ui/components/default/braveToday/cards/CardImage.tsx#L85

I suppose this could be done in rust before the item is returned.

If it's any consolation, our f/e (via React) won't inject these strings as html. You have to call `_dangerouslySetInnerHTML` which we don't.

Reply:
Ah I see. It could be done in Rust, or it could potentially be done right here in C++ so that you can set `metadata->image_url` (or similar) in addition to `metadata->description`. It's not a blocker, but if it were me, I'd probably do it like that so that I can avoid storing HTML that will never be used.

So what happen if there are HTML tags in the description (let's say `<b>`)? Does that get removed (or escaped?) in React?

Reply:
Description field is now escaped (and also not used by c++ anymore since we don't yet display it in any UI on any platform).

As for a " in an attribute, I suppose it might be possible if it were encoded in the source as `&quot;` or `&#34` but it wouldn't do much since 1) React would handle it correctly, 2) as would android I believe and 3) We don't (and cannot) set this url directly on an html `<img` - we have to ask c++ to download the image bytes for us - so the GURL gets sent back to the C++ which performs the download.